### PR TITLE
html_editor: introduce new html editor addon

### PR DESCRIPTION
**Pull request by ged-odoo**
This commit introduces a new implementation of the html editor. It is
intended to replace the web_editor implementation as soon as we can
remove all its current uses.

The new architecture is designed to be
- smaller: the scope of this addon is now only focused on edition, the
  Editor class and the Wysiwyg component. The massmailing
  sidebar/snippets system stays in web_editor
- extensible, via a plugin architecture, where each user of the editor
  can pick and choose which plugin/feature is going to be used in its
  specific usecase. Also, new plugin can be easily added to manage
  specific needs
- maintainable: each plugin is smallish and focus on one concern. For
   example, the selection plugin manages all things selection. This
   should make it easier to understand.

This addon is the result of a collaboration between the web editor team
and the javascript framework team. Many people contributed in various
ways to this project. Thank to everyone involved!

- Co-authored-by: abd-msyukyu-odoo <abd@odoo.com>
- Co-authored-by: Antoine Guenet <age@odoo.com>
- Co-authored-by: fdardenne <dafl@odoo.com>
- Co-authored-by: David Monjoie <dmo@odoo.com>
- Co-authored-by: FrancoisGe <fge@odoo.com>
- Co-authored-by: Géry Debongnie <ged@odoo.com>
- Co-authored-by: Jinjiu Liu <jili@odoo.com>
- Co-authored-by: Lucas Perais <lpe@odoo.com>
- Co-authored-by: Nicolas Bayet <nby@odoo.com>
- Co-authored-by: Rodolpho Lima <rcdl@odoo.com>
- Co-authored-by: Sébastien Geelen (sge) <sge@odoo.com>

### CODE REVIEW

**Comment by rfr-odoo** [comment]
@robodoo override=ci/style
Inline CSS Style (in JS file)

**Comment by ged-odoo** [comment]
Dear @odoo/rd-security 
Here is a massive PR to introduce a new html_editor addon in community. The goal is to replace web_editor as soon as we can adapt all the code that relies on it. Security wise, it is red, which is to be expected from a html editor. I investigated all flags and provide you with my rationale on why it is safe. Could you validate that this is okay? Thank you!

# Preliminary security review

## addons/html_editor/static/lib/DOMpurify.js:770
this is library code, same as web_editor

## addons/html_editor/static/src/core/clipboard_plugin.js

l245: the html is parsed into a fragment and immediately sanitized (at the next line)
l339: same

## addons/html_editor/static/src/editor.js:97

This is the main entry point for the editor. It is the place where we insert
the html that we want to edit in the DOM. This is assumed to be trusted.


## addons/html_editor/static/src/fields/html_viewer.js:73

The HTML Viewer component is a sub component used by the field html to display
html when the field is in readonly. The code for the html field assume that the
content is trusted (same as before).

## addons/html_editor/static/src/main/banner_plugin.js:52

The parseHTML call is there to generate a html element for the banner. The content
of the banner is mostly static, and all dynamic elements are actually hardcoded
in the file above, so at the end, this is safe (assuming that we trust translations)

## addons/html_editor/static/src/main/font/font_plugin.js:277

The replaceWith method here is used to replace a useless htmlelement by a new
`<p>` paragraphelement. Since the `p` value is just hardcoded one line above the
flagged line, it is safe

## addons/html_editor/static/src/main/list/list_plugin.js:338

Here, we transform `<li>` into `<p>` if they are not in a `<ul>` or `<ol>`. The
paragraph element is hardcoded a line above the flagged line, so it is safe

## addons/html_editor/static/src/main/list/list_plugin.js:387

The code in this function convert a P in a LI/SPAN structure. Looking at the 
three lines above, we can see that it is only manipulating the DOM, but does not
insert any dynamic/untrusted content

## addons/html_editor/static/src/main/media/media_plugin.js:135

The code here is in the callback function called by the media dialog whenever
the user has completed editing a media element. The `element` here is the value
sent by the media dialog (which should be safe), and the node is the targetted
value in the DOM. This line here replace the node by the updated value from the
media dialog. It seems to me that it is necessary to the user workflow.

## addons/html_editor/static/src/main/signature_plugin.js:32

The code here inserts the user signature. It is the result of a web_read operation
on the "signature" field. Looking at the few lines above the parseHTML operation,
it seems ok to me. Also, it is the exact translation of current code (see
`static/src/js/wysiwyg/wysiwyg.js`, l2479)

## addons/html_editor/static/src/main/table/table_plugin.js:120

The parsed html here is generated just a few line above. It is the html for a
new html table, which does not contain dynamic content.

#	addons/html_editor/static/src/others/chatgpt/chatgpt_dialog.js:69

The string that is markuped is generated from the content of a chat gpt api call.
However, a few line above the markup, the code is sanitizing the content in place.

#	addons/html_editor/static/src/others/collaboration/collaboration_odoo_plugin.js:597

This code runs when the editor needs to reset values from the server in collaboration
mode. The content here has to be trusted, otherwise the collaboration process
cannot work

## addons/html_editor/static/src/others/star_plugin.js:74

The content that is parsed here is hardcoded just the line above (it is a sequence
of stars)

## addons/html_editor/static/tests/*

All security issues flagged here are in the test helpers and in some cases,
hardcoded values in a few tests. They are only executed in the unit test
environment.

## addons/html_editor_playground/static/src/card.js:18

The innerHTML call is done in a temporary demo addon (html_playground) that will
not be merged. It is here for now to make it easy to test specific features, but
will be removed before final merge.

## addons/mail/static/tests/inline/convert_inline.test.js:26

This is done in a test. Should be ok.

**Review by mart-e — APPROVED** [review]
as mentionned by xavier, just update the dompurify version please

File: addons/html_editor/static/lib/DOMpurify.js
Comment (ged-odoo) [reply addons/html_editor/static/lib/DOMpurify.js:1]: well, the real reason is that it's just a move from the web_editor addon

**Comment by ged-odoo** [comment]
@mart-e thanks. will work on updating dompurify right away

**Comment by ged-odoo** [comment]
@mart-e as requested, dompurify has been updated (see commit https://github.com/odoo/odoo/pull/149028/commits/c6ddda1757c50d04f654c1d96807dd19ef9ece71)

**Comment by ged-odoo** [comment]
robodoo rebase-merge

**Comment by mart-e** [comment]
@robodoo override=ci/security
Reason: https://github.com/odoo/odoo/pull/149028#issuecomment-2200240486

**Comment by ged-odoo** [comment]
robodoo r+

**Comment by ged-odoo** [comment]
robodoo merge

**Comment by ged-odoo** [comment]
robodoo retry

**Pull request merged**
